python/ccf/clients.py

```diff
@@ -294,23 +294,39 @@ class RequestClient:
     CCF default client and wrapper around Python Requests, handling HTTP signatures.
     """
 
-    def __init__(self, host, port, ca, cert=None, key=None):
+    def __init__(
+        self,
+        host: str,
+        port: int,
+        ca: str,
+        cert: Optional[str] = None,
+        key: Optional[str] = None,
+    ):
         self.host = host
         self.port = port
         self.ca = ca
         self.cert = cert
         self.key = key
         self.session = requests.Session()
         self.session.verify = self.ca
-        self.session.cert = (self.cert, self.key)
+        if self.cert is not None and self.key is not None:
+            self.session.cert = (self.cert, self.key)
         self.session.mount("https://", TlsAdapter(self.ca))
 
-    def request(self, request, signed=False, timeout=DEFAULT_REQUEST_TIMEOUT_SEC):
+    def request(
+        self,
+        request: Request,
+        signed: bool = False,
+        timeout: int = DEFAULT_REQUEST_TIMEOUT_SEC,
+    ):
         extra_headers = {}
         extra_headers.update(request.headers)
 
         auth_value = None
         if signed:
+            if self.key is None:
```

Comment:
We ought to remove the signed flag and pass the key explicitly, and sign with that key when it's provided. That could be different from self.key, legitimately.

Reply:
I don't think that CCF supports that - the HTTP signature is verified using the caller's TLS certificate.

Reply:
Looking at the code now, no, but it was definitely the idea that proxying/sending requests on behalf of would work that way.

Reply:
We discussed a solution for this, but haven't implemented it. The problem is key identification, finding the key the signer claims was used, that we should verify against. The simplest way is to use the caller's. But we can use the `keyId` in the HTTP signature scheme, set to something like `user0` or `member1`, and then verify proxy'd signed requests.